python/ray/train/lightning/lightning_predictor.py

```diff
@@ -0,0 +1,143 @@
+            to prediction.
+        use_gpu: If set, the model will be moved to GPU on instantiation and
+            prediction happens on GPU.
+    """
+
+    def __init__(
+        self,
+        model: pl.LightningModule,
+        preprocessor: Optional["Preprocessor"] = None,
+        use_gpu: bool = False,
+    ):
+        self.model = model
+        self.model.eval()
+        self.use_gpu = use_gpu
+
+        self.device = torch.device("cpu")
+        if use_gpu:
+            # TODO (jiaodong): #26249 Use multiple GPU devices with sharded input
+            self.device = torch.device("cuda")
+
+        # Ensure input tensor and model live on the same device
+        self.model.to(self.device)
+
+        if (
+            not use_gpu
+            and torch.cuda.device_count() > 0
+            and log_once("lightning_predictor_not_using_gpu")
+        ):
+            logger.warning(
+                "You have `use_gpu` as False but there are "
+                f"{torch.cuda.device_count()} GPUs detected on host where "
+                "prediction will only use CPU. Please consider explicitly "
+                "setting `LightningPredictor(use_gpu=True)` or "
+                "`batch_predictor.predict(ds, num_gpus_per_worker=1)` to "
+                "enable GPU prediction."
+            )
+
+        super().__init__(preprocessor)
+
+    def __repr__(self):
+        return (
+            f"{self.__class__.__name__}(model={self.model!r}, "
+            f"preprocessor={self._preprocessor!r}, use_gpu={self.use_gpu!r})"
+        )
+
+    @classmethod
+    def from_checkpoint(
+        cls,
+        checkpoint: Checkpoint,
+        model: pl.LightningModule,
+        model_init_config: Dict["str", Any] = {},
+        preprocessor: Optional[Preprocessor] = None,
+        use_gpu: bool = False,
+    ) -> "LightningPredictor":
+        checkpoint = LightningCheckpoint.from_checkpoint(checkpoint)
+        model = checkpoint.get_model(model, model_init_config)
+        return cls(model=model, preprocessor=preprocessor, use_gpu=use_gpu)
+
+    @DeveloperAPI
+    def call_model(
+        self, inputs: Union[torch.Tensor, Dict[str, torch.Tensor]]
+    ) -> Union[torch.Tensor, Dict[str, torch.Tensor]]:
+        """Runs inference on a single batch of tensor data.
+
+        This method is called by `LightningPredictor.predict` after converting the
+        original data batch to torch tensors.
+
+        Override this method to add custom logic for processing the model input or
+        output.
+
+        Args:
+            inputs: A batch of data to predict on, represented as either a single
+                PyTorch tensor or for multi-input models, a dictionary of tensors.
+
+        Returns:
+            The model outputs, either as a single tensor or a dictionary of tensors.
+
+        """
+        with torch.no_grad():
+            # TODO(yunxuanx): How to get the batch_idx value?
+            output = self.model.predict_step(inputs, batch_idx=0)
```

Comment:
Actually, I hesitate choosing between `self.model.forward()` and `self.model.predict_step()`. 

If the users didn't define `predict_step()` it will call `forward()` by default. The benefit is the users can define some post processing logic in `predict_step()`, such as taking argmax over the output probabilities.

The only flaw here is that we cannot pass `batch_idx` for the user.

Reply:
yeah I'm not sure if `predict_step` is meant to be called directly given that we don't know what batch_idx is.

let's just start with doing vanilla torch prediction first.

In fact, if LightningCheckpoint subclasses TorchCheckpoint, then we don't need to create a LightningPredictor at all.